docs/design-go-api.md

```diff
@@ -0,0 +1,218 @@
+# Go API Design documentation
+
+## Overview
+
+This document presents the high-level user API for the Go-Wrapper client library. Specifically, it demonstrates how the user connects to Redis, executes requests, receives responses, and checks for errors.
+
+## Requirements
+
+- The minimum supported Go version will be 1.18. This version was chosen because it added support for generics, including type constraints and type sets.
+- The API will be thread-safe.
+- The API will accept as inputs all of [RESP2 types](https://github.com/redis/redis-specifications/blob/master/protocol/RESP2.md). We plan to add support for RESP3 types when they are available.
+- The API will attempt authentication, topology refreshes, reconnections, etc., automatically. In case of failures concrete errors will be returned to the user.
```

Comment:
These 3 requirements are copy-pasted from design-api.md, I'm assuming they are accurate for Go.